include/syscall.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2017, Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache 2.0
+ */
+
+
+#ifndef _ZEPHYR_SYSCALL_H_
+#define _ZEPHYR_SYSCALL_H_
+
+/* Fixed system call IDs. We use #defines instead of enumeration so that if
+ * system calls are retired it does not shift the IDs of other system calls.
+ */
+#define K_SYSCALL_BAD 0
+
+#define K_SYSCALL_LIMIT	1
+
+#ifndef _ASMLANGUAGE
+#include <misc/printk.h>
+
+/**
+ * @typedef _k_syscall_handler_t
+ * @brief System call handler function type
+ *
+ * These are kernel-side skeleton functions for system calls. They are
+ * necessary to sanitize the arguments passed into the system call:
+ *
+ * - Any kernel object or device pointers are validated with _SYSCALL_IS_OBJ()
+ * - Any memory buffers passed in are checked to ensure that the calling thread
+ *   actually has access to them
+ * - Many kernel calls do no sanity checking of parameters other than
+ *   assertions. The handler must check all of these conditions using
+ *   _SYSCALL_ASSERT()
+ * - If the system call has more then 5 arguments, then arg5 will be a pointer
+ *   to some struct containing arguments 5+. The struct itself needs to be
+ *   validated like any other buffer passed in from userspace, and its members
+ *   individually validated (if necessary) and then passed to the real
+ *   implementation like normal arguments
+ *
+ * Even if the system call implementation has no return value, these always
+ * return something, even 0, to prevent register leakage to userspace.
+ *
+ * Once everything has been validated, the real implementation will be executed.
+ *
+ * @param arg1 system call argument 1
+ * @param arg2 system call argument 2
+ * @param arg3 system call argument 3
+ * @param arg4 system call argument 4
+ * @param arg5 system call argument 5
+ * @param ssf System call stack frame pointer. Used to generate kernel oops
+ *            via _arch_syscall_oops_at(). Contents are arch-specific.
+ * @return system call return value, or 0 if the system call implementation
+ *         return void
+ *
+ */
+typedef u32_t (*_k_syscall_handler_t)(u32_t arg1, u32_t arg2, u32_t arg3,
+				      u32_t arg4, u32_t arg5, void *ssf);
+
+
+extern const _k_syscall_handler_t _k_syscall_table[K_SYSCALL_LIMIT];
+
+#define _SYSCALL_ASSERT(expr, ssf) \
+	do { \
+		if (!(expr)) { \
+			printk("system call failed: " #expr "\n"); \
```

Comment:
Clarity: "system call failed" implies that this is an application logging a routine runtime failure when it's actually a fatal error.

Naming: calling this "ASSERT" implies to the reader that it's debug code that will be removed in an optimized build and potentially removable when maintaining code, when it's actually a critical piece of the security model.  Suggest "SYSCALL_VALIDATE" or somesuch.

Reply:
sure I'll rename it